### PR TITLE
Add patches fields for AKS resources

### LINKED ISSUE #3629: Automate AKS features available in ASO
/kind feature

**Describe the solution you'd like**
As a user of CAPZ I would like the ability to provision all features available in AKS as soon as they are made available.  

Even once we switch to the dependency of using ASO for AKS provisioning, all features present in ASO will take individual coding work to be able to be utilized in CAPZ.  Since the features in AKS rapidly update ([see comment](https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2625#issuecomment-1557652425) on #2625), it would be ideal to have a mechanism for people to access all the available options in ASO via CAPZ via some coding mechanism which auto generates the code for it to work.

### PR DESCRIPTION
**Pull request by nojnhuh**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**: This implements #4304 for AzureManagedControlPlane by adding `spec.asoManagedClusterPatches` and for AzureManagedMachinePool by adding `spec.asoManagedClustersAgentPoolPatches` to allow users to specify arbitrary patches to apply to those ASO resources.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3629

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [X] squashed commits
- [ ] includes documentation
- [X] adds unit tests

**Release note**:

```release-note
EXPERIMENTAL: Added fields to AzureManagedControlPlane and AzureManagedMachinePool to supply arbitrary JSON merge patches to underlying ASO resources
```

### CODE REVIEW

File: api/v1beta1/types_class.go [hunk 1]
```diff
@@ -225,6 +225,10 @@ type AzureManagedControlPlaneClassSpec struct {
 	// AutoUpgradeProfile defines the auto upgrade configuration.
 	// +optional
 	AutoUpgradeProfile *ManagedClusterAutoUpgradeProfile `json:"autoUpgradeProfile,omitempty"`
+
+	// ASOManagedClusterPatches defines JSON merge patches to be applied to the generated ASO ManagedCluster resource.
+	// +optional
+	ASOManagedClusterPatches []string `json:"asoManagedClusterPatches,omitempty"`
```
Comment (nojnhuh) [inline api/v1beta1/types_class.go:231]: TODO: add this to AzureManagedControlPlane

File: api/v1beta1/types_class.go [hunk 2]
```diff
@@ -235,6 +235,13 @@ type AzureManagedControlPlaneClassSpec struct {
 	// SecurityProfile defines the security profile for cluster.
 	// +optional
 	SecurityProfile *ManagedClusterSecurityProfile `json:"securityProfile,omitempty"`
+
+	// ASOManagedClusterPatches defines JSON merge patches to be applied to the generated ASO ManagedCluster resource.
+	// WARNING: This is meant to be used sparingly to enable features for development and testing that are not
+	// otherwise represented in the CAPZ API. Misconfiguration that conflicts with CAPZ's normal mode of
+	// operation is possible.
```
Comment (nojnhuh) [inline api/v1beta1/types_class.go:242]: @dtzar Feel free to offer feedback on how scary this message should be. I also tried to reflect this sentiment in the release note in the PR description.

**Comment by nojnhuh** [comment]
This is ready for review but we still need to wait for consensus on the proposal first.
/hold for #4304

File: api/v1beta1/types_class.go (hunk 2 above)
Comment (dtzar) [reply api/v1beta1/types_class.go:242]: LGTM

File: test/e2e/aks_patches.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+			defer GinkgoRecover()
+			defer wg.Done()
+
+			ammp := &infrav1.AzureManagedMachinePool{}
+			Expect(mgmtClient.Get(ctx, types.NamespacedName{
+				Namespace: mp.Spec.Template.Spec.InfrastructureRef.Namespace,
+				Name:      mp.Spec.Template.Spec.InfrastructureRef.Name,
+			}, ammp)).To(Succeed())
+
+			nonAdditionalTagKeys := map[string]struct{}{}
```
Comment (willie-yao) [inline test/e2e/aks_patches.go:133]: Where is this variable used? I only see it being set on line 138

File: azure/services/aso/aso.go [hunk 4]
```diff
@@ -220,11 +226,64 @@ func (r *reconciler[T]) CreateOrUpdateResource(ctx context.Context, spec azure.A
...
+	parameters.SetGroupVersionKind(gvk)
+	paramData, err := json.Marshal(parameters)
+	if err != nil {
+		return zero, errors.Wrap(err, "failed to marshal JSON for patch")
+	}
+
+	for i, extraPatch := range p.ExtraPatches() {
+		jsonPatch, err := yaml.ToJSON([]byte(extraPatch))
+		if err != nil {
+			return zero, errors.Wrapf(err, "failed to convert patch at index %d to JSON", i)
```
Comment (willie-yao) [inline azure/services/aso/aso.go:264]: Can we print the patch here as well for more clarity?

File: azure/services/aso/aso.go [hunk 5]
```diff
@@ -220,11 +226,64 @@ func (r *reconciler[T]) CreateOrUpdateResource(ctx context.Context, spec azure.A
...
+		}
+		paramData, err = jsonpatch.MergePatch(paramData, jsonPatch)
+		if err != nil {
+			return zero, errors.Wrapf(err, "failed to apply patch at index %d", i)
```
Comment (willie-yao) [inline azure/services/aso/aso.go:268]: Same as above

**Review by willie-yao — COMMENTED** [review]
Looks good! Just a few minor non-blocking comments from my end.

File: azure/services/aso/aso.go (hunk 4 above)
Comment (nojnhuh) [reply azure/services/aso/aso.go:264]: These patches could be long or contain newlines, so I'm not sure if that would help much here. I think the index gives enough info to let people know where to look.

**Review by willie-yao — APPROVED** [review]
/lgtm

**Comment by nojnhuh** [comment]
So is the test just flaky? 🧐 

/test pull-cluster-api-provider-azure-e2e-aks

**Comment by Jont828** [comment]
Looks like all tests are green! Feel free to remove the hold whenever you're ready to merge

/lgtm
/approve

**Comment by Jont828** [comment]
Actually, please squash your commits first, LGTM otherwise.

/approve cancel

File: api/v1beta1/types_class.go [hunk 6]
```diff
@@ -235,6 +235,13 @@ type AzureManagedControlPlaneClassSpec struct {
...
+	// +optional
+	ASOManagedClusterPatches []string `json:"asoManagedClusterPatches,omitempty"`
```
Comment (jackfrancis) [inline api/v1beta1/types_class.go:244]: I have a slight preference for `ASOPatches` or `ServiceOperatorPatches` as a name.

I know that the CAPZ type name isn't always the same as the equivalent ASO resource name that we use (AzureManagedControlPlane vs ManagedCluster) but I think that level of detail is a UX annoyance.

Comment (nojnhuh) [reply api/v1beta1/types_class.go:244]: I was trying to make sure we're not backed into a corder API-wise by naming this something generic if it really means something more specific. If we ever let users patch other kinds of resources like the vnet that comes with AzureManagedControlPlane, it wouldn't be clear what something called `ASOPatches` is operating against. And given this field is meant to be a bit of a hack, I think this tradeoff is worth not having ideal UX.

Comment (dtzar) [reply api/v1beta1/types_class.go:244]: I agree with Jon here.  If we get to where there ever is a more general ASO patch, then we can name the new field accordingly at that time.

**Review by dtzar — APPROVED** [review]

**Comment by dtzar** [comment]
Not blocking, but a nice to have would be some kind of addition in the docs of how to patch (e.g. use asoctl to export/compare).  If not in this PR, then would be good to add a new docs issue accordingly.

File: api/v1beta1/types_class.go (hunk 6 above)
Comment (jackfrancis) [reply api/v1beta1/types_class.go:244]: sgtm!

**Comment by nojnhuh** [comment]
I'll plan to document this as part of #4568.

**Review by mboersma — APPROVED** [review]
/lgtm
/approve

File: api/v1beta1/types_class.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mboersma) [reply api/v1beta1/types_class.go:242]: PR still needs squash.

**Comment by nojnhuh** [comment]
squashed!
/hold cancel

**Pull request merged**